packages/flutter/lib/src/foundation/basic_types.dart

```diff
@@ -280,3 +280,37 @@ class _LazyListIterator<E> implements Iterator<E> {
     return true;
   }
 }
+
+/// A factory interface, that also reports the type of the created objects.
+///
+/// The equality of [Factory] object is determined solely by [Factory.type].
+class Factory<T> {
+  /// Creates a new factory.
+  ///
+  /// If `type` is null, [Factory.type] returns [T], else it returns `type`.
+  ///
+  /// The `constructor` parameter must not be null.
+  const Factory(this.constructor, {Type type})
```

Comment:
I'm concerned about the `type` argument, especially because (as far as I can tell) we have no way to verify that it's correct. When I saw this API, I thought "that's going to be a recipe for bugs", then I searched for "type:", and the first example in this PR is, as far as I can tell, an example of a bug.

I don't really understand why we need to override the `type` though? Isn't a `Factory< VerticalDragGestureRecognizer>` a subclass of `Factory<OneSequenceGestureRecognizer>` and thus capable of being put in the same `List<Factory<OneSequenceGestureRecognizer>>` list?